src/jit/copyprop.cpp

```diff
@@ -180,6 +180,13 @@ void Compiler::optCopyProp(BasicBlock* block, GenTreeStmt* stmt, GenTree* tree,
             continue;
         }
 
+        // Do not copy propagate if the old lclVar was 'doNotEnregister', but not the new one, in case
+        // this is an address-taken context.
+        if (lvaTable[lclNum].lvDoNotEnregister && !lvaTable[newLclNum].lvDoNotEnregister)
```

Comment:
It's not very clear what the comment is trying to say. What is this address-taken context it is talking about and why if it's address-taken does the code check for `lvDoNoEnregister` and not `lvAddrExposed`? Well, I guess checking for `lvAddrExposed` would be pointless since those variables won't be tracked anyway. But it's still not clear what this issue is about.

Reply:
Sorry this isn't clear; I'll try to figure out how to word it better. Here's the output from what it was doing:
```
N010 (  7,  5) [001151] n---G-------                       +--*  IND       double $849
N009 (  3,  3) [000766] ------------                       |  \--*  ADDR      byref  $253
N008 (  3,  2) [000764] -------N----                       |     \--*  LCL_VAR   simd12 V128 tmp110      u:2 $142
...
VN based copy assertion for [000764] V128 @00000142 by [001130] V91 @00000142.
N008 (  3,  2) [000764] -------N----              *  LCL_VAR   simd12 V128 tmp110      u:2 $142
copy propagated to:
N008 (  3,  2) [000764] -------N----              *  LCL_VAR   simd12 V91 tmp73
```
In this case, V91 is `lvDoNotEnregister` but V128 is not. The assert then happens in liveness because we're trying to take the address of a variable that's potentially going to be in a register.
I was kind of surprised that copyprop will modify nodes marked `GTF_DONT_CSE`, but I think that's reasonable, as long as we don't have this case.

In my comment, I was trying to express that I'm using the `lvDoNotEnregister` specifically to avoid this case.

I initially tried disabling copyprop to nodes marked `GTF_DONT_CSE`, but as I feared this had more regressions than one might like. With that change, for frameworks & benchmarks on x64/windows I had:

Total bytes of diff: 6198 (0.01% of base)
907 total methods with size differences (46 improved, 861 regressed), 249655 unchanged.

With this change we have:

Total bytes of diff: -64 (0.00% of base)
513 total methods with size differences (124 improved, 389 regressed), 250049 unchanged.

Reply:
Ah, another case of pesky reinterpretation that leaves variables DNER not address exposed. It's interesting that this `IND(ADDR(LCL_VAR))` wasn't flattened to a `LCL_FLD`, that would have avoided the problem.

> I was kind of surprised that copyprop will modify nodes marked GTF_DONT_CSE, but I think that's reasonable, as long as we don't have this case.

Hmm, I think the really surprising part is that we copy prop something that's not actually a value but a location (since it's under `GT_ADDR`). I wonder if this actually works properly...

> Sorry this isn't clear; I'll try to figure out how to word it better. Here's the output from what it was doing:

Including the IR dump above should be enough IMO.

Reply:
> In this case, V91 is lvDoNotEnregister but V128 is not. The assert then happens in liveness because we're trying to take the address of a variable that's potentially going to be in a register.

I am confused, in your example V91 has `lvDoNotEnregister` set, V128 doesn't have it. We replace V128 with V91, so the old var (V128) had ``lvDoNotEnregister` and the new one doesn't (V91).

`if (lvaTable[91].lvDoNotEnregister <- this is false && !lvaTable[128].lvDoNotEnregister <- this is false)`

It looks like 128 has `lvDoNotEnregister` and 91 doesn't, is it correct?

Also, do we want to replace a var that is not address taken with address taken? I think in some cases such restriction will hurt performance, but I am not sure if we have enough information to make this decision in `optCopyProp`.

Reply:
> It looks like 128 has lvDoNotEnregister and 91 doesn't, is it correct?

Yes, I stated it backward.

> Also, do we want to replace a var that is not address taken with address taken? I think in some cases such restriction will hurt performance, but I am not sure if we have enough information to make this decision in optCopyProp.

I thought perhaps the condition should be:
```
if (lvaTable[lclNum].lvDoNotEnregister != lvaTable[newLclNum].lvDoNotEnregister)
```
And the diffs were actually identical. I thought perhaps it was better to restrict only the incorrect case, but perhaps the above would be better.